common/lib/xmodule/xmodule/vertical_block.py

```diff
@@ -64,7 +64,11 @@ def _student_or_public_view(self, context, view):
                     usage_key=self.location),  # pylint: disable=no-member
             if 'username' not in child_context:
                 user_service = self.runtime.service(self, 'user')
-                child_context['username'] = user_service.get_current_user().opt_attrs['edx-platform.username']
+                xblock_user = user_service.get_current_user()
+                if 'edx-platform.username' in xblock_user.opt_attrs:
+                    child_context['username'] = xblock_user.opt_attrs['edx-platform.username']
+                else:
+                    view = PUBLIC_VIEW
```

Comment:
Yeah, your understanding is correct!

I think what you're suggesting would be ideal, but I'm not familiar enough with the xmodule code to know how to achieve that. I was having a hard time tracking down what code path brings it to VerticalBlock.student_view() rather than VerticalBlock.public_view(), despite setting PUBLIC_VIEW in the sequential 😕 

`_student_or_public_view()` is called by both VerticalBlock.public_view() and VerticalBlock.student_view(), the only difference between the two being the view parameter that's passed in... so it might not be the _worst_ to leave this switch to PUBLIC_VIEW here in this method...

What are your thoughts?

Reply:
@ciduarte I'm surprised that setting PUBLIC_VIEW in the sequential isn't doing it; feels like a bug. Taking a closer look at the seq_module now...

Reply:
Maybe the issue is that we hard-code `'student_view'` here? https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/views/views.py#L1662

Reply:
@ciduarte Yeah, I confirmed with `pdb` that line above is what's preventing `'public_view'` from making it down from the request to the Vertical render call.

> _student_or_public_view() is called by both VerticalBlock.public_view() and VerticalBlock.student_view(), the only difference between the two being the view parameter that's passed in... so it might not be the worst to leave this switch to PUBLIC_VIEW here in this method...

I would argue that switching the provided view type based on a condition in the runtime environment is bit of a bamboozle. Like you pointed out, the code in `common/lib/xmodule/` is already tough to follow, so it's probably in our best interest to minimize special cases. Fortunately, we also have a blended project going currently to simplify it.

Thanks for doing the due diligence of reaching out about this code. Hopefully the fix is as simple as rendering`view` instead of `'student_view'` in that courseware API view I linked.

Reply:
I appreciate you digging in! I'll clean this up and try that line. Thank you so much 😁